velox/functions/prestosql/aggregates/AverageAggregate.cpp

```diff
@@ -279,19 +309,98 @@ class AverageAggregate : public exec::Aggregate {
       } else {
         clearNull(rawNulls, i);
         auto* sumCount = accumulator(group);
-        rawValues[i] = TResult(sumCount->sum) / sumCount->count;
+        if constexpr (
+            (std::is_same_v<TResult, double> || std::is_same_v<TResult, float>)&&(
+                std::is_same_v<TAccumulator, double> ||
+                std::is_same_v<TAccumulator, float>)) {
+          rawValues[i] = TResult(sumCount->sum) / sumCount->count;
+        } else if constexpr ((std::is_same_v<TResult, UnscaledLongDecimal> ||
+                              std::is_same_v<
+                                  TResult,
+                                  UnscaledShortDecimal>)&&std::
+                                 is_same_v<TAccumulator, int128_t>) {
+          // Need to perform integer division here as the intermediate sums may
+          // overflow decimal limits.
+          int128_t result = 0;
+          DecimalUtil::divideWithRoundUp<int128_t, int128_t, int64_t>(
+              result, sumCount->sum, sumCount->count, 0, 0);
+          rawValues[i] = TResult(result);
+        }
       }
     }
   }
 
   DecodedVector decodedRaw_;
   DecodedVector decodedPartial_;
-};
+}; // class AverageAggregate
+
+/// Override 'accumulatorAlignmentSize' for UnscaledLongDecimal values as it
+/// uses int128_t type. Some CPUs don't support misaligned access to int128_t
+/// type.
+template <>
+inline int32_t
+AverageAggregate<UnscaledShortDecimal, int128_t, UnscaledShortDecimal>::
+    accumulatorAlignmentSize() const {
+  return static_cast<int32_t>(sizeof(int128_t));
+}
+
+template <>
+inline int32_t
+AverageAggregate<UnscaledLongDecimal, int128_t, UnscaledLongDecimal>::
+    accumulatorAlignmentSize() const {
+  return static_cast<int32_t>(sizeof(int128_t));
+}
+
+// This overload required for final aggregator after partially aggregating a
+// vector short decimals.
+template <>
+inline int32_t
+AverageAggregate<UnscaledLongDecimal, int128_t, UnscaledShortDecimal>::
+    accumulatorAlignmentSize() const {
+  return static_cast<int32_t>(sizeof(int128_t));
+}
+
+template <>
+template <bool tableHasNulls>
+void AverageAggregate<UnscaledShortDecimal, int128_t, UnscaledShortDecimal>::
+    updateNonNullValue(char* group, UnscaledShortDecimal value) {
```

Comment:
The body of the three specializations is the same. Probably use a single template function.